File: Introibo/Design/PageFrame.swift
import SwiftUI

// Two shared visual treatments that every Introibo page uses:
//   1. The gold hairline page frame  (PageFrameModifier)
//   2. The subtle paper grain overlay (PaperGrainOverlay)
//
// Apply both with .pageChrome() — or use them individually.

struct PageFrameModifier: ViewModifier {
    var inset: CGFloat = 10

    func body(content: Content) -> some View {
        content
            .overlay(
                ZStack {
                    // Outer hairline
                    Rectangle()
                        .stroke(Color.frameLine, lineWidth: 0.5)
                    // Faint inner hairline, 3pt in
                    Rectangle()
                        .stroke(Color.frameLine.opacity(0.5), lineWidth: 0.5)
                        .padding(3)
                }
                .padding(inset)
                .allowsHitTesting(false)
            )
    }
}

struct PaperGrainOverlay: ViewModifier {
    func body(content: Content) -> some View {
        content
            .overlay(
                // Subtle dot pattern using a Canvas — three layered radial
                // dot fields at different scales, multiplied on top of the
                // page to add tactile warmth without washing out the content.
                Canvas { ctx, size in
                    let layers: [(step: CGFloat, radius: CGFloat, alpha: Double, offset: CGPoint)] = [
                        (step: 7,  radius: 0.6, alpha: 0.025, offset: CGPoint(x: 1, y: 2)),
                        (step: 13, radius: 0.7, alpha: 0.02,  offset: CGPoint(x: 5, y: 9)),
                        (step: 11, radius: 0.6, alpha: 0.018, offset: CGPoint(x: 3, y: 7)),
                    ]
                    let dotColor = Color(red: 92/255, green: 60/255, blue: 30/255)
                    for layer in layers {
                        var y: CGFloat = layer.offset.y
                        while y < size.height {
                            var x: CGFloat = layer.offset.x
                            while x < size.width {
                                let rect = CGRect(
                                    x: x - layer.radius, y: y - layer.radius,
                                    width: layer.radius * 2, height: layer.radius * 2
                                )
                                ctx.fill(Path(ellipseIn: rect), with: .color(dotColor.opacity(layer.alpha)))
                                x += layer.step
                            }
                            y += layer.step
                        }
                    }
                }
                .blendMode(.multiply)
                .opacity(0.7)
                .allowsHitTesting(false)
            )
    }
}

struct PageChromeModifier: ViewModifier {
    func body(content: Content) -> some View {
        content
            .background(Color.pageBackground.ignoresSafeArea())
            .modifier(PaperGrainOverlay())
            .modifier(PageFrameModifier())
    }
}

extension View {
    /// Applies the full Introibo page look: parchment background (walnut
    /// in dark mode) + paper grain overlay + gold hairline frame.
    /// Use as the top-level modifier on each screen's root container.
    func pageChrome() -> some View {
        modifier(PageChromeModifier())
    }

    /// Just the gold hairline frame, without the paper grain. For dark
    /// headers or overlays that have their own background treatment.
    func pageFrame(inset: CGFloat = 10) -> some View {
        modifier(PageFrameModifier(inset: inset))
    }
}

File: android/app/src/main/java/com/lampstandhq/introibo/ui/components/PageFrame.kt
package com.lampstandhq.introibo.ui.components

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.BlendMode
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.lampstandhq.introibo.ui.theme.IntroiboTheme
import com.lampstandhq.introibo.ui.theme.RawPalette

// ---------------------------------------------------------------------------
// Paper grain overlay
// ---------------------------------------------------------------------------
// Subtle dot pattern drawn via Canvas — three layered radial dot fields at
// different scales, composited in multiply mode on top of the page to add
// tactile warmth without washing out content.  Mirrors the iOS
// PaperGrainOverlay in PageFrame.swift.

/**
 * Three-layer grain dot specification.
 */
private data class GrainLayer(
    val step: Float,
    val radius: Float,
    val alpha: Float,
    val offsetX: Float,
    val offsetY: Float,
)

private val grainLayers = listOf(
    GrainLayer(step = 7f,  radius = 0.6f, alpha = 0.025f, offsetX = 1f, offsetY = 2f),
    GrainLayer(step = 13f, radius = 0.7f, alpha = 0.02f,  offsetX = 5f, offsetY = 9f),
    GrainLayer(step = 11f, radius = 0.6f, alpha = 0.018f, offsetX = 3f, offsetY = 7f),
)

private fun DrawScope.drawGrainOverlay() {
    val dotBaseColor = RawPalette.GrainDot
    for (layer in grainLayers) {
        val dotColor = dotBaseColor.copy(alpha = layer.alpha * 0.7f)
        var y = layer.offsetY
        while (y < size.height) {
            var x = layer.offsetX
            while (x < size.width) {
                drawCircle(
                    color = dotColor,
                    radius = layer.radius,
                    center = Offset(x, y),
                    blendMode = BlendMode.Multiply,
                )
                x += layer.step
            }
            y += layer.step
        }
    }
}

// ---------------------------------------------------------------------------
// Gold frame border overlay
// ---------------------------------------------------------------------------
// Two concentric hairline rectangles in the gold-leaf frame colour, matching
// the iOS PageFrameModifier.

private fun DrawScope.drawGoldFrame(frameLineColor: Color, inset: Float) {
    // Outer hairline
    drawRect(
        color = frameLineColor,
        topLeft = Offset(inset, inset),
        size = Size(size.width - inset * 2, size.height - inset * 2),
        style = Stroke(width = 0.5f),
    )
    // Inner hairline, 3dp further in
    val inner = inset + 3f
    drawRect(
        color = frameLineColor.copy(alpha = frameLineColor.alpha * 0.5f),
        topLeft = Offset(inner, inner),
        size = Size(size.width - inner * 2, size.height - inner * 2),
        style = Stroke(width = 0.5f),
    )
}

// ---------------------------------------------------------------------------
// Public composables
// ---------------------------------------------------------------------------

/**
 * Paper grain overlay — add as a sibling inside a [Box] to layer the grain
 * on top of content, or wrap content with [PageChrome] for the full look.
 */
@Composable
fun PaperGrainOverlay(modifier: Modifier = Modifier) {
    Canvas(modifier = modifier.fillMaxSize()) {
        drawGrainOverlay()
    }
}

/**
 * Gold hairline page frame overlay.
 *
 * @param inset Distance from the edges for the outer hairline (default 10dp).
 */
@Composable
fun PageFrameOverlay(
    inset: Dp = 10.dp,
    modifier: Modifier = Modifier,
) {
    val frameLineColor = IntroiboTheme.colors.frameLine
    Canvas(modifier = modifier.fillMaxSize()) {
        drawGoldFrame(frameLineColor, inset.toPx())
    }
}

/**
 * Full Introibo page chrome: parchment/walnut background + paper grain
 * overlay + gold hairline frame. Use as the root container of each screen.
 *
 * This is the Compose equivalent of the iOS `.pageChrome()` modifier.
 *
 * ```
 * PageChrome {
 *     LazyColumn { ... }
 * }
 * ```
 */
@Composable
fun PageChrome(
    modifier: Modifier = Modifier,
    inset: Dp = 10.dp,
    content: @Composable () -> Unit,
) {
    val colors = IntroiboTheme.colors

    Box(
        modifier = modifier
            .fillMaxSize()
            .background(colors.pageBackground),
    ) {
        // Screen content
        content()

        // Paper grain layer (multiply blend)
        PaperGrainOverlay()

        // Gold frame on top
        PageFrameOverlay(inset = inset)
    }
}
